### PR TITLE
New provider: Mail.ru

### PR DESCRIPTION
**Pull request by justdmitry**
Please add new provider: **Mail.ru**

Homepage: https://mail.ru

OAuth docs: http://api.mail.ru/docs/guides/restapi/

### CODE REVIEW

**Comment by PinpointTownes** [comment]
Thanks for your contribution, it looks fine :+1: 

I'll merge it as soon as we're done with the `project.json` -> `.csproj` migration.

File: src/AspNet.Security.OAuth.MailRu/project.json [hunk 1]
```diff
@@ -0,0 +1,50 @@
+﻿{
+  "version": "1.0.0-beta3-*",
```
Comment (PinpointTownes) [inline src/AspNet.Security.OAuth.MailRu/project.json]: Note for me: `rc1`.

File: src/AspNet.Security.OAuth.MailRu/MailRuAuthenticationHandler.cs [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+        protected override async Task<AuthenticationTicket> CreateTicketAsync([NotNull] ClaimsIdentity identity,
+            [NotNull] AuthenticationProperties properties, [NotNull] OAuthTokenResponse tokens)
+        {
+            // Mail.ru REST API rules:
+            //   1. method name (users.getInfo) is passed as param
+            //   2. MD5 hash computed over all params (ordered alphabetically),
+            //        concatenated 'name1=value1name2=value2...signkey' without separators (like &)
+            //   3. MD5 Hash value passed as 'sig' param
+
+            var requestParams = "app_id=" + Options.ClientId + "&format=json&method=users.getInfo&secure=1&session_key=" + tokens.AccessToken;
```
Comment (PinpointTownes) [inline src/AspNet.Security.OAuth.MailRu/MailRuAuthenticationHandler.cs]: Please consider moving that to a separate method. You can follow the same pattern as the one we used for the Instagram provider: https://github.com/aspnet-contrib/AspNet.Security.OAuth.Providers/blob/dev/src/AspNet.Security.OAuth.Instagram/InstagramAuthenticationHandler.cs#L77

Comment (PinpointTownes) [inline src/AspNet.Security.OAuth.MailRu/MailRuAuthenticationHandler.cs]: Please consider moving that to a separate method. You can follow the same pattern as the one we used for the Instagram provider: https://github.com/aspnet-contrib/AspNet.Security.OAuth.Providers/blob/dev/src/AspNet.Security.OAuth.Instagram/InstagramAuthenticationHandler.cs#L77

Comment (justdmitry) [inline src/AspNet.Security.OAuth.MailRu/MailRuAuthenticationHandler.cs]: Done.

File: src/AspNet.Security.OAuth.MailRu/MailRuAuthenticationHandler.cs [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+
+                return OAuthTokenResponse.Failed(new Exception("An error occurred while retrieving an access token."));
+            }
+
+            var payload = JObject.Parse(await response.Content.ReadAsStringAsync());
+
+            return OAuthTokenResponse.Success(payload);
+        }
+
+        protected virtual string BuildUrlWithSignature(string baseUrl, IDictionary<string, string> parameters, string signKey)
```
Comment (PinpointTownes) [inline src/AspNet.Security.OAuth.MailRu/MailRuAuthenticationHandler.cs]: Unfortunately, this method won't work if the user adds custom query string parameters to the `Options.UserInformationEndpoint` address. To fix that, you'll need to parse the address, just like what the Instagram provider does.

**Pull request closed without merging**